In [61]:
import pandas as pd
from time import time
from fasttext import supervised 

In [63]:
#Download dbpedia dataset from: 
#https://github.com/le-scientifique/torchDatasets/raw/master/dbpedia_csv.tar.gz
data_path = '/home/bangaru/Downloads/NLPBookTut/'
# Loading train data
train_file = data_path + 'dbpedia_csv/train.csv'
df = pd.read_csv(train_file, header=None, names=['class','name','description'])
# Loading test data
test_file = data_path + 'dbpedia_csv/test.csv'
df_test = pd.read_csv(test_file, header=None, names=['class','name','description'])
# Data we have
print("Train:{} Test:{}".format(df.shape,df_test.shape))


Train:(560000, 3) Test:(70000, 3)


In [64]:
# Since we have no clue about the classes lets build one
# Mapping from class number to class name
class_dict={
            1:'Company',
            2:'EducationalInstitution',
            3:'Artist',
            4:'Athlete',
            5:'OfficeHolder',
            6:'MeanOfTransportation',
            7:'Building',
            8:'NaturalPlace',
            9:'Village',
            10:'Animal',
            11:'Plant',
            12:'Album',
            13:'Film',
            14:'WrittenWork'
        }

# Mapping the classes
df['class_name'] = df['class'].map(class_dict)
df.head()

class                               name  \
0      1                   E. D. Abbott Ltd   
1      1                     Schwan-Stabilo   
2      1                         Q-workshop   
3      1  Marvell Software Solutions Israel   
4      1        Bergan Mercy Medical Center   

                                         description class_name  
0   Abbott of Farnham E D Abbott Limited was a Br...    Company  
1   Schwan-STABILO is a German maker of pens for ...    Company  
2   Q-workshop is a Polish company located in Poz...    Company  
3   Marvell Software Solutions Israel known as RA...    Company  
4   Bergan Mercy Medical Center is a hospital loc...    Company

In [65]:
df["class_name"].value_counts()

NaturalPlace              40000
Athlete                   40000
MeanOfTransportation      40000
Company                   40000
Animal                    40000
Album                     40000
Village                   40000
Plant                     40000
WrittenWork               40000
Film                      40000
OfficeHolder              40000
Building                  40000
EducationalInstitution    40000
Artist                    40000
Name: class_name, dtype: int64

In [66]:
# Lets do some cleaning of this text
def clean_it(text,normalize=True):
    # Replacing possible issues with data. We can add or reduce the replacemtent in this chain
    s = str(text).replace(',',' ').replace('"','').replace('\'',' \' ').replace('.',' . ').replace('(',' ( ').\
            replace(')',' ) ').replace('!',' ! ').replace('?',' ? ').replace(':',' ').replace(';',' ').lower()
    
    # normalizing / encoding the text
    if normalize:
        s = s.normalize('NFKD').str.encode('ascii','ignore').str.decode('utf-8')
    
    return s

# Now lets define a small function where we can use above cleaning on datasets
def clean_df(data, cleanit= False, shuffleit=False, encodeit=False, label_prefix='__class__'):
    # Defining the new data
    df = data[['name','description']].copy(deep=True)
    df['class'] = label_prefix + data['class'].astype(str) + ' '
    
    # cleaning it
    if cleanit:
        df['name'] = df['name'].apply(lambda x: clean_it(x,encodeit))
        df['description'] = df['description'].apply(lambda x: clean_it(x,encodeit))
    
    # shuffling it
    if shuffleit:
        df.sample(frac=1).reset_index(drop=True)
            
    return df

In [67]:
%%time
# Transform the datasets using the above clean functions
df_train_cleaned = clean_df(df, True, True)
df_test_cleaned = clean_df(df_test, True, True)

CPU times: user 4.44 s, sys: 68.1 ms, total: 4.51 s
Wall time: 4.51 s


In [68]:
# Write files to disk as fastText classifier API reads files from disk.
train_file = data_path + 'dbpedia_train.csv'
df_train_cleaned.to_csv(train_file, header=None, index=False, columns=['class','name','description'] )

test_file = data_path + 'dbpedia_test.csv'
df_test_cleaned.to_csv(test_file, header=None, index=False, columns=['class','name','description'] )


Now that we have the train and test files written into disk in a format fastText wants, we are ready to use it for text classification!

In [69]:
## Using fastText for feature extraction and training
from fasttext import supervised 
"""fastText expects and training file (csv), a model name as input arguments.
label_prefix refers to the prefix before label string in the dataset.
default is __label__. In our dataset, it is __class__. 
There are several other parameters which can be seen in: 
https://pypi.org/project/fasttext/
"""
%time model = supervised(train_file, 'temp', label_prefix="__class__")
results = model.test(test_file)
print(results.nexamples, results.precision, results.recall)

CPU times: user 56.5 s, sys: 1.51 s, total: 58 s
Wall time: 12.6 s
70000 0.9710571428571428 0.9710571428571428


Try training a classifier on this dataset with, say, LogisticRegression to realize how fast fastText is! 97% Precision and Recall are hard numbers to beat, too!